File: soln-array-problems/rotate-list.java
/*
https://leetcode.com/problems/rotate-list/
*/

import java.util.*;
import java.lang.*;
import java.io.*;

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if(head==null || head.next==null)
            return head;
        
        int count=1;
        ListNode tail=head;
        while(tail.next!=null){
            tail=tail.next;
            count++;
        }
        tail.next=head;
        k=k%count;
        int i=1;
        ListNode p=head;

        while(i<(count-k)){
            i++;
            p=p.next;
        }
        head=p.next;
        p.next=null;
        return head;
    }
}

File: Python/soln-linkedlist-problems/rotate-list.py
'''
 @author Farheen Bano
  
 Reference-
 https://leetcode.com/problems/rotate-list/
'''

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if head is None or head.next is None:
            return head
        
        count=1
        tail=head
        while tail.next is not None:
            tail=tail.next
            count=count+1
        
        tail.next=head
        k=k%count
        i=1
        p=head

        while(i<(count-k)):
            i=i+1
            p=p.next
        
        head=p.next
        p.next=None
        return head